biodatasets/chemdner/chemdner.py

```diff
@@ -0,0 +1,392 @@
+                    "offsets": [
+                        (
+                            loc.offset - eo,
+                            loc.offset + loc.length - eo,
+                        )
+                        for loc in a.locations
+                    ],
+                    "text": (a.text,),
+                    "id": a.id,
+                    "normalized": self._get_normalized(a),
+                }
+
+                pe.append(da)
+
+            entities.append(pe)
+
+        return passages, entities
+
+    def _get_normalized(self, a: bioc.BioCAnnotation) -> List[Dict]:
+        """
+        Get normalization DB and ID from annotation identifiers
+        """
+
+        identifiers = a.infons.get("identifier")
+
+        if identifiers is not None:
+
+            identifiers = re.split(r",|;", identifiers)
+
+            identifiers = [i for i in identifiers if i != "-"]
+
+            normalized = [i.split(":") for i in identifiers]
+
+            normalized = [{"db_name": elems[0], "db_id": elems[1]} for elems in normalized]
+
+        else:
+
+            normalized = [{"db_name": "-1", "db_id": "-1"}]
+
+        return normalized
+
+    def _generate_examples(
+        self,
+        filepath: str,
+        split: str,
+    ) -> Iterator[Tuple[int, Dict]]:
+        """Yields examples as (key, example) tuples."""
+
+        reader = bioc.BioCXMLDocumentReader(str(filepath))
+
+        if self.config.schema == "source":
+
+            for uid, doc in enumerate(reader):
+
+                passages, passages_entities = self._get_passages_and_entities(doc)
+
+                if len(passages) < 1 or len(passages_entities) < 1 or len(passages_entities) != len(passages):
+                    continue
+
+                for p, pe in zip(passages, passages_entities):
+
+                    p.pop("offsets")  # BioC has only start for passages offsets
+
+                    p["document_id"] = doc.id
+                    p["entities"] = pe  # BioC has per passage entities
+
+                yield uid, {"passages": passages}
+
+        elif self.config.schema == "bigbio_kb":
+            uid = 0
+            for idx, doc in enumerate(reader):
+
+                passages, passages_entities = self._get_passages_and_entities(doc)
+
+                if len(passages) < 1 or len(passages_entities) < 1 or len(passages_entities) != len(passages):
+                    continue
+
+                # global id
+                uid += 1
+
+                # unpack per-passage entities
+                entities = [e for pe in passages_entities for e in pe]
+
+                for p in passages:
+                    p.pop("offset")  # drop original offset
+                    p["text"] = (p["text"],)  # text in passage is Sequence
+                    p["id"] = uid  # override BioC default id
+                    uid += 1
+
+                for e in entities:
+                    e["id"] = uid  # override BioC default id

```

Comment:
Please make sure here that entities without normalization (i.e. with `{"db_name": "-1", "db_id": "-1"}`) have `e["normalized"] = []` instead of `e["normalized"] = [{"db_name": "-1", "db_id": "-1"}]` for the BigBio schema.